javalin/src/main/java/io/javalin/json/JsonMapper.kt

```diff
@@ -28,6 +30,29 @@ interface JsonMapper {
      */
     fun toJsonStream(obj: Any, type: Type): InputStream = throw NotImplementedError("JsonMapper#toJsonStream not implemented")
 
+    /**
+     * Javalin uses this method for [io.javalin.http.Context.writeJsonStream]. When implementing this method
+     * you are expected to consume `stream` one element at a time, convert that element to JSON, and write that
+     * element's corresponding JSON to the `outputStream`. The intent is to reduce the amount of memory used
+     * during transformation of a data stream to JSON.
+     *
+     * It is implied that the `stream` of objects will be transformed to an Array type in JSON. In other words,
+     * it is your responsibility to surround the output with `[` and `]`.
+     *
+     * Use `ctx.outputStream()` for the compressed output stream. Or use `ctx.res().outputStream` if you need the
+     * raw uncompressed stream.
+     *
+     * You are responsible for setting the content type header, which is likely to be "application/json" done
+     * by calling `ctx.contentType(ContentType.APPLICATION_JSON)`.
+     *
+     * Once Javalin calls this function, it will not participate any further in the response. It becomes
+     * your responsibility to deliver the remainder of the response to the `outputStream`.
+     *
+     * When your implementation returns from the `writeStream` function, assume that the underlying resources
+     * for the `stream` will be released.
+     */
+    fun writeStream(ctx: Context, stream: Stream<*>): Unit = throw NotImplementedError("JsonMapper#writeStream not implemented")
```

Comment:
@tipsy and @dzikoysk  Should we pass in ctx: Context? or pass in outputStream: OutputStream?

Passing in `ctx` gives future flexibility to whoever implements their own JsonMapper. They may want to use a JSON content type variant, or use an uncompressed stream for some reason.

If we pass in `outputStream` instead, the docs would say that content type is already set to `application/json`, and that the output stream is compressed. Doing this would be a more opinionated stance on what we expect this function to do.

I'm looking for feedback before writing some unit tests.

Reply:
I didn't read full conversation about this feature, but in this particular case imo we shouldn't, Context is dependent on JsonMapper, but JsonMapper is not on Context, especially that we're considering JsonMapper as general usage abstraction layer for any json implementation user can use in any Javalin component. Such mapper should provide quite generic interface to make it possible to handle such use-case, but necessary impl that fully cover http requirements should be in Context/Context util.

Reply:
> Context is dependent on JsonMapper, but JsonMapper is not on Context

Sounds good, I'll pass in an `outputStream` instead and write some unit tests accordingly. Thanks for the helpful feedback.